### DISCUSSION TITLE
Support flatpak

### DISCUSSION THREAD

**Discussion by baimafeima** [mullvad/mullvadvpn-app#8901] [Feature requests & Ideas]
The best solution is obviously to have `mullvadvpvn-app` included in the repositories of an operating system. But as this will take lots of time and people use different operating systems, could you make an additional flatpak release for easier cross-platform installation? This would also allow to have it hosted on [Flathub](https://flathub.org/home) and attract more potential users. When making a Flatpak release, keep in mind that the [security](https://flatkill.org/) will depend on how well it is packaged by the vendor.

**Comment by thomhoess** [comment]
Please! I'm on Solus and it's a hassle to get up and running, if not Flatpak maybe Appimage (#1123 ) or Snap (#847 ) - one of them would be sufficient, it would be great to have it in an format in which it can run everywhere - my personal taste would say to make a flatpak. Thank you!

**Comment by encryptophile** [comment]
+1 for this. I've been a Mullvad user years, and strongly support the project. Offering a universial installation package (Flatpak, Appimage, etc) would be extremely helpful.

**Comment by joffaMac** [comment]
Please, make a flatpak of your great product. Please. Thank you.

**Comment by Ramblurr** [comment]
The Silverblue team at Fedora offered to help do this.. https://twitter.com/teamsilverblue/status/1273212527137165313

Did anyone at mullvad ever followup? The Silverblue team would be a great resource to tap to create a flatpak.

**Comment by joffaMac** [comment]
Thanks for following up but alas no, Mullvad seem to have absolutely no interest in making a flatpak.  Back on 1Nov2020 I sent an email asking and got this reply:

     We currently don't offer any flatpack. We do only offer a rpm and .deb package currently.
     I will forward the information to our developers but I don't think we currently will add any flatpack.

I've done a couple of reply tweets to some of the Mullvad news updates since that email, trying to suggest they make a flatpak, but never had a comment/reply back.

As I'm just a user and not a developer, I've decided Mullvad clearly have no interest in flatpak, so I use the CLI on my 2 Fedora Silverblue machines (as per this url - https://blogs.gnome.org/thaller/2019/03/15/wireguard-in-networkmanager/) and their app on my Android phone.

**Comment by 1player** [comment]
Any update? This has been open for 3 years, and Flatpaks are here to stay. Can the community help in any way?

**Comment by joffaMac** [comment]
> Any update? This has been open for 3 years, and Flatpaks are here to stay. Can the community help in any way?

No update from me as I read somewhere recently that Mullvad don't have the resources/time to look at a flatpak version - if the community can help, then please go for it.

**Comment by poperigby** [comment]
Still no update?

**Comment by ghost** [comment]
> > Any update? This has been open for 3 years, and Flatpaks are here to stay. Can the community help in any way?
> 
> No update from me as I read somewhere recently that Mullvad don't have the resources/time to look at a flatpak version - if the community can help, then please go for it.



> Still no update?

Any update why not drop Snap one and do Flatpak instead ?
Snap is used exclusivly on ubuntu, so dropping it for flatpak (any distribution is better).

**Comment by faern** [comment]
We don't have any Snap support to drop. Any Snap package you find is a third party package. We only distribute the Linux app via the RPM and DEB files we upload to our website and Github. ANY other distribution method is an unofficial third party distribution of our app on Linux.

**Comment by Kabouik** [comment]
This is not a great long term solution, but you might be interested in this: #3412


https://user-images.githubusercontent.com/7107523/157864056-0fe96f81-499e-46e6-9c4e-d7eae0bbaddd.mp4

**Comment by ghost** [comment]
I wish for an official flatpak support as well. I've been using Mullvad for months, but I recently moved to Linux and decided to use Arch distro, and while the GUI client is available through AUR, it's not maintained by Mullvad, is it? 

The installation instructions for Linux on Mullvad website mentions AUR package here:
https://aur.archlinux.org/packages/mullvad-vpn-bin
The arch wiki mentions this one instead:
https://aur.archlinux.org/packages/mullvad-vpn

Both are maintained by the same maintainer.

Might be unrelated, but the package requires importing GPG key from Mullvad itself during installation, which enhances security. I have checked the PKGBUILDS, and while they don't have malicious code in them, does Mullvad trust this maintainer, and can I trust this them too? Having unofficial maintainer of a sensitive application gives me slight concern. Of course, I can always use wireguard, but I just want to be sure.

**Comment by faern** [comment]
> Arch distro, and while the GUI client is available through AUR, it's not maintained by Mullvad, is it?

Correct. The AUR packages are not maintained by us!

Regarding `mullvad-vpn-bin` vs `mullvad-vpn`. The former just unpacks a prebuilt package. The latter builds the app from source. If packaged correctly both should yield the same app.

> I have checked the PKGBUILDS, and while they don't have malicious code in them, does Mullvad trust this maintainer, and can I trust this them too?

We don't endorse the usage of this third party package. We have also not seen any malicious code in that package. But since we are not in control of the package we can't give any guarantees as to how it will behave in the future. You are correct to be concerned. If you want to be careful, check out the PKGBUILDS before each upgrade, or use vanilla WireGuard IMO.

**Comment by dginovker** [comment]
I've posted a Request on Flathub's Discourse, if everyone can go show support over there we might get the attention of someone in the community with the talent to do this :tada:! https://discourse.flathub.org/t/vpn-mullvad-vpn/2594

**Comment by who-biz** [comment]
@thomhoess 

> Please! I'm on Solus and it's a hassle to get up and running, if not Flatpak maybe Appimage (#1123 ) or Snap (#847 ) - one of them would be sufficient, it would be great to have it in an format in which it can run everywhere - my personal taste would say to make a flatpak. Thank you!

I have made an AppImage-compiling source available here: https://github.com/who-biz/mullvadvpn-app-appimage.  I am running it on Solus.  See README for additional information, and use these steps as support if you get confused: https://discuss.getsol.us/d/7853-mullvad-client-20215.

Converting the build config in `gui/tasks/distribution.js` should not be difficult if you'd like to modify it further to generate  a Flatpak instead.

**Comment by 1player** [comment]
That AppImage is cool, but sadly still requires modifying the system files, which defeats the purpose IMHO.

There doesn't seem to be a way of neatly packaging Mullvad because it's a monolithic package that both ships a GUI and all the system files, while the correct approach would be to, at least on Linux, have it split between the system daemon part that needs root privileges are *require* to be installed with your regular package manager, and the standalone GUI part that can be distributed as Flatpak, AppImage, etc.

One example of this in the Flatpak world is Piper, a GUI to control your mouse configuration. It depends on the ratbagd daemon, which *has* to be installed with apt/dnf/what have you, while the Flatpak only provides the unprivileged GUI.

@faern Are there any plans on splitting the application accordingly, instead of putting everything into a single package? In other words, splitting the daemon which needs root access, and the UI in two separate packages that can be distributed separately (with the latter being shipped as Flatpak, Snap, AppImage, etc.)

**Comment by who-biz** [comment]
For users of distributions with custom package management systems (non-rpm-based), above solves the "I can't run mullvad except through NetworkManager and OpenVPN"... That repo is not meant to be ready-to-distribute, but rather a *get-this-working*.

Killswitch, Wireguard, other things that do not exist or work poorly, in NetworkManager, are at least functional if compiled this way.

**Comment by kinghat** [comment]
ivpn has a snap package but forego split tunneling as a tradeoff: https://www.ivpn.net/apps-linux/#snap

**Comment by Kabouik** [comment]
For unsupported distributions using systemd, there is also https://github.com/mullvad/mullvadvpn-app/issues/3412. It is not a great solution, and it uses opinionated folders, but it works great for me on Solus with very little to do to install or update. I'm very glad to see an AppImage too, surely this will be used by a wider audience.

However, I have moved recently to Guix, which is more like NixOS than regular distributions, and there neither the installer script nor the AppImage would work. A Flatpak would work, though, so I'd be happy if Mullvad could release one one day. In the mean time, I use this: https://gitlab.com/emacsomancer/volemad

**Comment by faern** [comment]
> @faern Are there any plans on splitting the application accordingly, instead of putting everything into a single package? In other words, splitting the daemon which needs root access, and the UI in two separate packages that can be distributed separately (with the latter being shipped as Flatpak, Snap, AppImage, etc.)

We have been discussing it a little bit. But no concrete plans. We have discussed it only in terms of allowing users who prefer the CLI and/or want to use it on a headless server etc to install just the daemon + cli with a smaller package.

How is for example the solus problem solved by splitting the packages? As you write, you need to install the system service anyway, and that has to be done via the system package manager.

We have not research flatpak a whole lot. But it seems like the wrong type of package for this type of software. Flatpak is designed to isolate and sandbox your application. To protect your system from the application. Our software is kind of the other way around. We need to take control over the system to protect the user and other software. We need full access to the firewall, setting DNS, editing the routing table etc.

**Comment by 1player** [comment]
> Flatpak is designed to isolate and sandbox your application.

Yes and no. Flatpak is designed to simplify distribution. But your confusion is valid, as a VPN is a very particular application that requires privileged access to the host, so it doesn't map perfectly to this model.

The problem is that the regular distribution methods, such as deb/rpm/etc. have their own massive drawbacks as well, and the reason many distributions are moving to Flatpak is to reduce the maintenance and security burden of packaging and distributing everything via deb, but, if possible, to let the upstream developer package once and distribute to all Linux users, whatever their distro. Since it runs on a Linux namespace (i.e. container), it solves the "works on my machine" issue, as the only hard dependency is the kernel.

So I understand your reluctance to see how Flatpak solves the Mullvad problem. It doesn't, but solves many other problems with simpler GUI applications, and if this program were split into Flatpak GUI and deb/rpm privileged core, it would fit neatly in this "new world order", if you will :)

Also, the problem with the current monolitic rpm/deb is that immutable distributions (such as Silverblue, or even the Steam Deck OS), have a read only /, except /etc, /home and /var, and Mullvad's packages write stuff outside of these read-write mounts, which of course breaks. These distribution models are quite niche but they're the direction the Linux userspace is going, such as Fedora or indeed what Valve has chosen to use for the Deck. I use Silverblue on my workstation, and sadly mullvad is the only application I cannot install because it doesn't play nice with this read-only root model, so I had to create a script to connect manually via wireguard-tools.

Hope this answers some of your questions.

**Comment by faern** [comment]
How/where do you suggest we install the `mullvad-daemon` part of our application on a system with read-only root?

**Comment by 1player** [comment]
It can be installed anywhere in the read-only root (the system package manager has a way to access the read-only root as read-write when installing a package), but it should *not* write anything unless it's in /etc, /home or /var. 

Sorry, I forget the specifics since it's been a few months since I've tried the rpm on my machine, but IIRC mullvad creates a log file under its `/opt/Mullvad VPN` root and that causes problems on that kind of systems (https://github.com/mullvad/mullvadvpn-app/issues/1570). But that's a separate issue.

**Comment by faern** [comment]
Hmm. We keep our daemon logs in `/var/log/mullvad-vpn`. Maybe something with electron-builder causes unexpected files to be spread out? After install we really should not write to anything except `/etc/mullvad-vpn`, `/var/log/mullvad-vpn` and `/var/cache/mullvad-vpn`. Please file anything other than that as separate issue if you have information about it.

**Comment by the-spyke** [comment]
If Flatpak doesn't support something required for a VPN client, then it's a good reason to open a discussion on Flatpak/XDG if a new portal or protocol is required.

**Comment by CaptainMuskrat** [comment]
There is a strong need for a Flatpak and/or Appimage release. It's strange that the Mullvad browser has a flatpak yet the actual VPN it's supposed to complement does not.

**Comment by faern** [comment]
I'm going to cite my own post from earlier in this thread:

> We have not research flatpak a whole lot. But it seems like the wrong type of package for this type of software. Flatpak is designed to isolate and sandbox your application. To protect your system from the application. Our software is kind of the other way around. We need to take control over the system to protect the user and other software. We need full access to the firewall, setting DNS, editing the routing table etc.

The way our app is designed is that it takes control over the system (in certain areas). You as a user rely on us to guarantee security and anonomity. We feel like doing this from within a sandbox is backwards. I'm not saying Flatpak can't do this, we have not researched it yet. So it's bath a question of it feeling wrong, and a question of not had the time to investigate this yet.

**Comment by the-spyke** [comment]
@faern Even if it seems backward, Flatpak is the way to distribute apps today. And Flathub is the way to discover software.

BTW, Ubuntu will soon have Desktop Core edition where drivers and the kernel are distributed as snaps. So, a VPN client is not a problem. Your iOS app is already sandboxed.

**Comment by sonnyp** [comment]
Hello, I'm involved with GNOME, Flatpak and Flathub. 

We are interested in supporting Flatpak clients such as the Mullvad VPN app. I'd be happy to talk about the philosophy of Flatpak and the constraints/requirements.

You can contact me at `{my-github-username}@gnome.org`

**Comment by robke96** [comment]
+1 for Flatpak support!

**Comment by spipau** [comment]
Please make this happen, I'm annoyed by manually downloading .deb files to update my VPN and I don't want to add a new repository to apt

**Comment by sunjam** [comment]
Considering deb support has existed only for months, maybe calm it way down on feeling annoyed. Instead, be grateful.

**Comment by spipau** [comment]
@sunjam I am super calm, just want to keep my software updated in the most modern and easiest way. I am a paying customer and I just share my frustration with the company so they can improve their service. If this would an open source software free to use, then I'd be happy to compile it myself and be endlessly thankful without any frustration. Just as I am thankful for the great work of Mullvad but I see some room for improvement on the Linux side as .debs to be downloaded or custom repos are simply not the Linux way. Every time my Mullvad VPN statusbar icon has a yellow dot I have to manually download a deb file, then search online for the terminal command as I always forget it because Mullvad VPN is the only software I use which is handling their updates in such a way. Therefore, hope to see flatpak support very soon :slightly_smiling_face:

**Comment by Kabouik** [comment]
While the hassle of manually downloading .deb files at every update is real, I think the real issue is .deb files are not distro-agnostic. There are many Linux users left behind when the company just provides .deb files for Linux support and call it a day.  

Ideally, a distro-agnostic package, such as Flatpak, Guix, or even Appimage (though Appimages won't work on all distributions either) would be very welcome, as well as builds for x86_64 and aarch64 architectures.

**Comment by sunjam** [comment]
> Every time my Mullvad VPN statusbar icon has a yellow dot I have to manually download a deb file, then search online for the terminal command

Try adding the [repository to your sources list](https://mullvad.net/en/help/install-mullvad-app-linux).  That will make it part of the regular system update process on Debian, Ubuntu, Fedora.  This is will assuage the need for any searching or additional commands.

**Comment by Mosswool** [comment]
I would like to voice my support for a mullvad flatpak as well.

If the concern is that the platform is too containerized, it is my understanding that individual packages can disclose the the access they'll need beforehand, including firewall, network, etc., presumably anything a VPN client may need. I've even seen apps who's requirements simply list "full system access".

Having a flatpak would be a boon to many users on immutable linux distributions. Thank you very much for considering it.

**Comment by codenyte** [comment]
+1 for Flatpak, as I just tried to install Mullvad on Bazzite (Universal Blue/Fedora Atomic) and bricked my system. Not the fault of Mullvad, rpm-ostree isn't a particularly great solution, but that's why Flatpak is needed for the image-based future of Linux distributions.

**Comment by Kabouik** [comment]
Actually installing Mullvad's .deb package on Droidian (droidian.org, basically Deian onto the Android kernel) also soft-bricked the system. This was more than a year ago and I didn't mention it because I thought it was specific to that unusual OS, ut if that was observed on another one, maybe it's worth mentioning.  

I could fix the sysh1stem from a recovery shell, to manualy uninstall the package and revert all the changes it made.

**Comment by Dark-Thoughts** [comment]
> +1 for Flatpak, as I just tried to install Mullvad on Bazzite (Universal Blue/Fedora Atomic) and bricked my system. Not the fault of Mullvad, rpm-ostree isn't a particularly great solution, but that's why Flatpak is needed for the image-based future of Linux distributions.

How did you even install it? If I try to install the .rpm file it just gives me an error. I tried installing Mullvad through various DistroBox instances but also without any success, since they don't seem to work with the needed background service or something. I can only use the static wireguard option but that's very limiting with how many sites block the various Mullvad IPs, which usually require for a quick switch of the servers. I'm already contemplating to shell out the addition 5 bucks for Proton-VPN, since they seem to have a Flathub hosted client. :(

**Comment by codenyte** [comment]
> How did you even install it?

Just a simple `rpm-ostree install file.rpm` It initially appeared to have worked, but after rebooting, I realized that it broke my system.

> Proton-VPN, since they seem to have a Flathub hosted client

Just so you know, it's an unofficial package. And their Linux client is incredibly buggy, I randomly get logged out all the time, connections fail, the app crashes or the server list just disappears. It's borderline unusable on all the distros I tried (Arch, Artix, Fedora, Fedora Silverblue, openSUSE)

I just switched back to regular Fedora, since Silverblue is to annoying and restrictive for me

**Comment by eylenburg** [comment]
> Just so you know, it's an unofficial package.

Not for much longer though: https://github.com/flathub/com.protonvpn.www/issues/349

**Comment by Dark-Thoughts** [comment]
>Just a simple rpm-ostree install file.rpm It initially appeared to have worked, but after rebooting, I realized that it broke my system.

That's exactly what errors out for me, but I guess that's a good thing if it indeed breaks the system. I've heard layering installs is generally not recommended though.

>I just switched back to regular Fedora, since Silverblue is to annoying and restrictive for me

Tried Fedora KDE before Bazzite and using the provided guides to install the media codecs broke the system for me. lol
I get the frustration too though. I just don't know what Distro to even try at this point because a lot of them either break or have issues straight out of the box and I guess the immutable nature of Bazzite causes it to already have a somewhat secure system backup function.

**Comment by randshell** [comment]
I appreciate the Mullvad team added the rpm repository actually. I think it's great to not have to do the package install manually and also have everything updated on a new release.

As for the immutable distros, at least for me, I had no issues layering the app from the Mullvad repo on Fedora Kinoite. In fact, it's been all working smoothly since an year ago when the repo was online.

**Comment by ZethiaZ** [comment]
+1 for this request as well. I understand flatpaks are a bit complicated for VPNs but at least SurfShark has an official Linux flatpak now too so it can be done. This would be a game changer for me as I could use my favorite distro instead of having to stick to something supported by mullvad.

**Comment by iMonZ** [comment]
> +1 for this request as well. I understand flatpaks are a bit complicated for VPNs but at least SurfShark has an official Linux flatpak now too so it can be done. This would be a game changer for me as I could use my favorite distro instead of having to stick to something supported by mullvad.

Definitely a +1

**Comment by bertob** [comment]
Something to note is that there are immutable OSes that are not based on Silverblue, with no package layering systems. GNOME OS is one example, but I suspect there will be many more as adoption of systemd's update model with A/B partitions instead of OStree accelerates.

For those cases you really need Flatpaks, as the only alternative is manually creating a sysext from the RPM package, or similar horrible hacks.

**Comment by Duplicake-fyi** [comment]
Please Mullvad team, make a flatpak! I'm on alpine Linux on one of my computers and I want Mullvad on it, but it's not supported.

Doing this would also mean they wouldn't need to maintain 2 separate repos.

**Comment by TheAnnoyingFruit** [comment]
I think the decision here is quite obvious considering there are vpns available as flatpak so it is absolutely doable. You either maintain packages for the distros you choose and people will always complain as you can’t possibly properly support all of them. Or you make a flatpak that by just existing supports all distros that support flatpaks. Even if it costs a little bit more development time the benefits here far outweigh that and you have been offered help doing so by people in the community.

**Comment by GuyDebord** [comment]
Wow it has been more than 7 years and nothing has been done about it. This is just sad. Others have done it so it can be done and it works. At this point there is no excuse. The problem is not Flatpak, it's simply a lack of will.

**Reply by Ramblurr** [reply]
I want flatpak support as much as you, but come on, don't be an ass. Github issues aren't the place to take out your anger.

**Reply by GuyDebord** [reply]
I merely stated facts without anger, only disappointment. You're the only one name calling here. If I was a developer with a lot of free time then maybe I would do it but it's not the case. Mullvad is a for profit company, not a small FLOSS project maintained by a developer on his free time out of pure goodwill. Expecting them to do it in 7 years isn't outrageous entitlement.

**Reply by Duplicake-fyi** [reply]
Decided to switch to Proton VPN because it's in my proton family plan, the flatpak is working well for me. Will probably resubscribe to mullvad when this gets finished

**Reply by sraaronrock** [reply]
Building VPNs inside the Flatpak sandbox is notoriously complex due to strict network and permission isolation, which breaks most VPN clients out of the box. I’ve been experimenting with packaging Mullvad VPN as a Flatpak, but the sandboxing model (lack of CAP_NET_ADMIN, restricted /dev/net/tun access, and D-Bus limitations) makes it a real challenge...

If anyone is interested in helping tackle these hurdles, my project is here: https://github.com/sraaronrock/net.mullvad.mullvadvpn

**Reply by Dark-Thoughts** [reply]
> Decided to switch to Proton VPN because it's in my proton family plan, the flatpak is working well for me. Will probably resubscribe to mullvad when this gets finished

I wish you luck. I got talked into it and had a 1 year subscription with the black friday deal and had nothing but issues with their VPN and their absolute garbage app. Constantly loses its login information, had to basically log into my PW manager first to copy & paste the password to be secured, which is just ridiculous. It would also often not connect on the first launch after boot and had to be closed & re-opened - every single time. It would also often lose connection, or the servers went down to a crawl in speed. The only good thing was that you could favorite some servers for quick access from the tray.

Gonna try Mozilla VPN for a year, since it's still not reasonably possible to get Mullvad to run on Bazzite. Unfortunately their browser addon isn't compatible with Linux and they don't have a way to quickly change servers. Hopefully something they work on...

**Reply by eylenburg** [reply]
> Gonna try Mozilla VPN for a year, since it's still not reasonably possible to get Mullvad to run on Bazzite. 

it is possible, it's just not elegant to install: https://gist.github.com/eylenburg/ee5593ef7daff417a82d1c31c362c984

On other immutable systems like SteamOS or KDE Linux it's a different story

**Reply by Dark-Thoughts** [reply]
> it is possible, it's just not elegant to install: https://gist.github.com/eylenburg/ee5593ef7daff417a82d1c31c362c984

In the past I tried layering the .rpm they provide on their website, but it would just error out. Another user at the time reported that layering worked for them, but then their system wouldn't boot anymore. Various resources also keep stressing that layering should be the last resort / avoided, which makes me shy away from attempting it again.

**Reply by randshell** [reply]
If anyone would like to try again, since new Fedora / Bazzite versions were released in the meantime, there's always an easy way to do a rollback to the previous working state, if anything goes wrong. I've added the Mullvad repo on my Fedora Kinoite and layered the app with no issues since a couple of years.

I think that, aside from edge cases like switching to Gnome from a Kinoite (KDE) system, layering and doing a rollback should be painless and not breaking. Layering is not necessarily the last resort. How would you go installing VirtualBox without it? If it makes you feel more at ease, this is my list of layered packages:

```
State: idle
Deployments:
  fedora:fedora/43/x86_64/kinoite
                  Version: 43.20251105.0 (2025-11-05T01:00:56Z)
               BaseCommit: b53ce02c638cad1f550904f8379e6c302a2ada4f5b5fd66d71fb3871c1aca414
             GPGSignature: Valid signature by C6E7F081CF80E13146676E88829B606631645531
                     Diff: 12 upgraded
      RemovedBasePackages: libavdevice-free libavfilter-free libavformat-free ffmpeg-free libpostproc-free libswresample-free libavutil-free libavcodec-free libswscale-free 7.1.2-2.fc43
          LayeredPackages: bleachbit blender bpftool distrobox ffmpeg gamescope hashcat kernel-devel-matched kmod-VirtualBox
                           libavcodec-freeworld libayatana-appindicator-gtk3 mullvad-vpn pipx plasma-wallpapers-dynamic
                           python3-libguestfs realtime-setup rocm-clinfo rocm-hip rocm-opencl rocm-runtime rocminfo
                           rpmfusion-free-release rpmfusion-nonfree-release scrcpy syncthing virt-manager virtiofsd VirtualBox zsh
            LocalPackages: akmods-keys-0.0.2-8.fc37.noarch opensnitch-1.7.2-1.x86_64 opensnitch-ui-1.7.2-1.noarch
                           proton-mail-1.10.0-1.x86_64
```

I think the most problems I had was with the rpmfusion repository when using VirtualBox or Steam, as sometimes the dependencies and versions get out of sync with the official Fedora repo and you need to wait a few days for the updates to work again.

Ultimately it's your choice, these are my 2c if there's of any help to you. As usual, the disclaimer is to **have proper backups** if anything goes wrong.

> Another user at the time reported that layering worked for them, but then their system wouldn't boot anymore

Perhaps there was some issue with the `mullvad-early-boot-blocking.service` SystemD service?

**Reply by Dark-Thoughts** [reply]
> Layering is not necessarily the last resort. How would you go installing VirtualBox without it?

I don't know. I ended up using Virtual Machine Manager since they have flatpaks, but I only really need it for Windows exclusive firmware apps anyway. Had to layer Java for jDownloader 2 though, since I could not get that to run exclusively in a distrobox. But all the documentary sources and other sites say so. `Only as a last resort [layer the VPN with rpm-ostree](https://docs.bazzite.gg/Installing_and_Managing_Software/rpm-ostree/)`
https://docs.bazzite.gg/General/VPN/

> Perhaps there was some issue with the `mullvad-early-boot-blocking.service` SystemD service?

I couldn't tell you. There was not much documentation or discussion, let alone actual help, and it's been quite a while since. Maybe I'll try again in a year when my current payment runs out and if there's still no flatpak available then.

**Reply by TheAnnoyingFruit** [reply]
> I don't know. I ended up using Virtual Machine Manager since they have flatpaks, but I only really need it for Windows exclusive firmware apps anyway. Had to layer Java for jDownloader 2 though, since I could not get that to run exclusively in a distrobox. But all the documentary sources and other sites say so. `Only as a last resort [layer the VPN with rpm-ostree](https://docs.bazzite.gg/Installing_and_Managing_Software/rpm-ostree/)`
> https://docs.bazzite.gg/General/VPN/

Fully unrelated but there is a jdownloader flatpak

**Reply by Dark-Thoughts** [reply]
> Fully unrelated but there is a jdownloader flatpak

Yes, but it says it is 15 years old.

**Reply by TheAnnoyingFruit** [reply]
> > Fully unrelated but there is a jdownloader flatpak
> 
> Yes, but it says it is 15 years old.

It’s wrong if you check the flatpak manifest it has been updated multiple times and if you install it you will see it is not out of date

**Reply by sraaronrock** [reply]
> Building VPNs inside the Flatpak sandbox is notoriously complex due to strict network and permission isolation, which breaks most VPN clients out of the box. I’ve been experimenting with packaging Mullvad VPN as a Flatpak, but the sandboxing model (lack of CAP_NET_ADMIN, restricted /dev/net/tun access, and D-Bus limitations) makes it a real challenge...
> 
> If anyone is interested in helping tackle these hurdles, my project is here: https://github.com/sraaronrock/net.mullvad.mullvadvpn

I was working on this project (https://github.com/sraaronrock/net.mullvad.mullvadvpn), but I couldn’t get it to work. There are too many limitations in the way Mullvad is architected and how Flatpak sandboxes applications, making both essentially incompatible 😵

**Comment by jonathancross** [comment]
**Note:** Mozilla VPN is [available via Flatpak](https://github.com/mozilla-mobile/mozilla-vpn-client/blob/91f7f92271e34b980bdff4b93109fdc76196b69b/docs/Building/linux.md?plain=1#L59) and [supports advanced features like split-tunneling](https://github.com/mozilla-mobile/mozilla-vpn-client/blob/91f7f92271e34b980bdff4b93109fdc76196b69b/docs/releasenotes/v2.4.md?plain=1#L3) + WireGuard.

Maybe copy some of their config?

**Reply by TheAnnoyingFruit** [reply]
Being a subscriber it seems that the split tunneling feature is not equally supported. Depending on your environment they disable it. On all my systems I do not have the option to use it. Unsure if that is cause of flatpak or my distro? (https://github.com/mozilla-mobile/mozilla-vpn-client/issues/10720) (https://github.com/mozilla-mobile/mozilla-vpn-client/issues/6394). The only other feature that is missing I would say is quantum resistant tunneling.

**Reply by faern** [reply]
~@TheAnnoyingFruit If you have Mullvad VPN installed on the same computer as Mozilla VPN, they disable their split tunneling. Because they took our split tunneling driver, and only one instance of it can be active at any one point in time. I have no idea how their flatpak setup is made, but I suspect that the split tunneling difference you are seeing is not because of flatpak.~

Scratch that. That is of course only relevant on Windows, and flatpak only relevant on Linux. I had a minor brainfart.

**Comment by Android789515** [comment]
Proton VPN is available as a Flatpak and works fine. Please make a Flatpak version.

**Reply by Eggroley** [reply]
ProtonVPN on flatpak doesn't have split-tunneling so unless Mullvad can figure it out, there will be limitations on that front. 

I'd still like a flatpak client regardless of whether or not it has split tunneling though.

**Comment by jntdofe** [comment]
It would be nice if someone talked to the Flatpak team about this issue. They recently said they are working on the limitations of the current design and features and are preparing for Flatpak Next (or Flatpak2).

If they are made aware of this limitation, this could probably be fixed, if not in the current Flatpak, then in Flatpak Next.
